packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts

```diff
@@ -43,6 +43,53 @@ export interface UserPoolConfig {
     readonly defaultAction?: UserPoolDefaultAction;
 }
 
+function isUserPoolConfig(obj: unknown): obj is UserPoolConfig {
+    return (obj as UserPoolConfig).userPool !== undefined;
+}
+
+/**
+ * Configuration for API Key authorization in AppSync
+ */
+export interface ApiKeyConfig {
+    /**
+     * Unique description of the API key
+     */
+    readonly apiKeyDesc: string;
+
+    /**
+     * The time from creation time after which the API key expires, using RFC3339 representation.
+     * It must be a minimum of 1 day and a maximum of 365 days from date of creation.
+     * Rounded down to the nearest hour.
+     * @default - 7 days from creation time
+     */
+    readonly expires?: string;
+}
+
+function isApiKeyConfig(obj: unknown): obj is ApiKeyConfig {
+    return (obj as ApiKeyConfig).apiKeyDesc !== undefined;
+}
+
+type AuthModes = UserPoolConfig | ApiKeyConfig;
```

Comment:
JSII doesn't support unions currently. Usually the workaround is to make an interface that multiple classes implement.

Reply:
I didn't get any compiler errors and I see examples of union types elsewhere.

I tried a common interface and I got errors complaining about them not being used in the exported classes, but I can try again.

Reply:
With

```ts
interface UserPoolConfig extends AuthMode { ... }
interface ApiKeyConfig extends AuthMode { ... }

export interface AuthorizationConfig {
    readonly defaultAuthorization?: AuthMode;
    readonly additionalAuthorizationModes?: [AuthMode]
}
```

I get:

```
error: [awslint:no-unused-type:@aws-cdk/aws-appsync.UserPoolDefaultAction] type or enum is not used by exported classes (declared at lib/graphqlapi.ts:17)
```

Because now the exported classes deal in terms of `AuthMode`.

This feels like a bug in `awslint` because I'm still using those types in a function (e.g., `userPoolDescFrom(upConfig: UserPoolConfig)`), but I'm unsure how to proceed.

Reply:
It packaged fine with the union type. In Java, it mapped to 

```java
@software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
public Builder defaultAuthorization(software.amazon.awscdk.services.appsync.UserPoolConfig defaultAuthorization) {
  this.defaultAuthorization = defaultAuthorization;
  return this;
}

@software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
public Builder defaultAuthorization(software.amazon.awscdk.services.appsync.ApiKeyConfig defaultAuthorization) {
  this.defaultAuthorization = defaultAuthorization;
  return this;
}
```